mypy/expandtype.py

```diff
@@ -251,9 +251,9 @@ def visit_type_var(self, t: TypeVarType) -> Type:
             if (tvar_id := repl.id) in self.recursive_tvar_guard:
                 return self.recursive_tvar_guard[tvar_id] or repl
             self.recursive_tvar_guard[tvar_id] = None
-            repl = repl.accept(self)
-            if isinstance(repl, TypeVarType):
-                repl.default = repl.default.accept(self)
+            repl.default = repl.default.accept(self)
```

Comment:
Types should be _never_ modified in place. I know you didn't introduce it, but this should be fixed ASAP. (I think there is also a chance some of the problems you try to fix may be caused by this).

Reply:
So the problem seems to be with getting the correct ID's. In the test case, we have `class D[T1=str, T2=T1]`.

```python
>>> self.variables 
{1: T1`604 = builtins.str, 2: T2`605 = T1`1 = builtins.str}
>>> self.recursive_tvar_guard
{604: T1`604 = builtins.str, 605: None}
>>> repl.default = repl.default.accept(self)
>>> self.variables 
{1: T1`604 = builtins.str, 2: T2`605 = T1`604 = builtins.str}
>>> self.recursive_tvar_guard
{604: T1`604 = builtins.str, 605: None}
```

Modifying `repl.default` in-place ensures that in the `ExpandTypeVisitor.variables` dictionary, the default of the dependent type variable gets updated to the new ID.